### PR TITLE
ARROW-5258: [C++/Python] Collect file metadata of dataset pieces

### PR DESCRIPTION
**Pull request by pearu**
This PR supersedes PR #4166 and provides
- [x] file metadata API (C++)
- [x] ParquetWriter `metadata` property, only available after calling `close` method (Python)
- [x] FileMetaData `to_dict` method for collecting file metadata information to a dictionary (Python)
- [x] support for `metadata_collector` kw argument to `ParquetWriter` and `write_to_dataset` for collecting the file metadata instances of dataset pieces.
- [x] unit-test

### CODE REVIEW

**Comment by pearu** [comment]
@wesm , please review.

File: cpp/src/parquet/file_writer.cc [hunk 1]
```diff
@@ -310,6 +311,7 @@ class FileSerializer : public ParquetFileWriter::Contents {
         num_row_groups_(0),
         num_rows_(0),
         metadata_(FileMetaDataBuilder::Make(&schema_, properties, key_value_metadata)) {
+    file_metadata_.reset();
```
Comment (wesm) [inline cpp/src/parquet/file_writer.cc]: Is this needed (it should be null at initialization time)?

File: cpp/src/parquet/file_writer.cc [hunk 2]
```diff
@@ -296,6 +296,7 @@ class FileSerializer : public ParquetFileWriter::Contents {
       Close();
     } catch (...) {
     }
+    file_metadata_.reset();
```
Comment (wesm) [inline cpp/src/parquet/file_writer.cc]: Is this needed?

**Review by wesm — COMMENTED** [review]
Interesting solution to the original problem. Is there not a desire to produce the `_metadata` file at all? Maybe we should create a new JIRA issue since this exposes all the file metadata from the pieces to the caller of `write_to_dataset`

It seems like the Appveyor failure is the result of a flake

Since a lot of this Python code is destined for the graveyard (being replaced by a portable C++ implementation -- which will still have bindings of course) we should be cautious about exposing a ton of new APIs where we have to worry about backwards compatibility. Perhaps mark this API as "experimental" so we don't end up with cranky users if we have to provide this functionality in some other way

**Comment by pearu** [comment]
@mrocklin, @martindurant, could you also review this pull request for ARROW-1983? In particular, I have the following questions (cc: @wesm ):
1. The `write_to_dataset` interface for collecting file metadata instances of a dataset is exemplified in [test_dataset_metadata](https://github.com/apache/arrow/pull/4236/files#diff-348d66d69aec7aedf0a6f753613acb43R2625). Would this be an approach that would be suitable for dask?
2. After collecting all file metadata of dataset pieces (possibly from different dask processes), what would be desired the interface for writing and reading the dataset metadata instances? Here's an initial proposal for Python:
```
def write_dataset_metadata(metadata_list, where):
    ...

def read_dataset_metadata(where, memory_map=False):
    ...
    return metadata_list
```
3. Clearly, each file metadata instance in `metadata_list` contains same information as in all metadata instances. Would it make sense and use to introduce a new `DatasetMetadata` class (in C++, also exposed to Python) that would collect the metadata information of dataset pieces in a more compact way as well as would provide I/O methods? (This also addresses the question 2.)

**Comment by wesm** [comment]
@pearu what do you think about having `write_to_dataset` append the actual FileMetaData objects instead of converting them to a dict? Then we can punt the file writing until a follow up patch

**Comment by pearu** [comment]
@wesm, the `metadata_list` contains the actual `FileMetaData` objects. The `to_dict` was introduced for convenience of tests as well as quickly viewing the information in `FileMetaData` objects.

**Comment by wesm** [comment]
Got it, thanks. Can you rebase to trigger a new build (to make sure the appveyor failure was a flake)? I'm OK to merge this with a new JIRA id (since the writing of the _metadata file will have to be addressed separately)

**Comment by pearu** [comment]
OK, I'll trigger the new build by getting rid of `file_metadata_.reset();` statements (that are not really needed with the current control flow).

File: cpp/src/parquet/file_writer.cc (hunk 1 above)
Comment (pearu) [reply cpp/src/parquet/file_writer.cc]: Removed.

File: cpp/src/parquet/file_writer.cc (hunk 2 above)
Comment (pearu) [reply cpp/src/parquet/file_writer.cc]: No, fixed.

**Review by wesm — APPROVED** [review]
+1, thanks @pearu for your patience on this, hopefully this will work for Dask

**Comment by rjzamora** [comment]
Thanks for your work on this @pearu and @wesm !  

I am trying to help @mrocklin improve the dask/parquet interface.  To this end, I put together a simple notebook to explore how `pyarrow.parquet` can/should be used to read/write a partitioned dataset without dask (see: https://github.com/rjzamora/notebooks/blob/master/pandas_pyarrow_simple.ipynb). 

In the current implementation, is there a way to read/write a single metadata file for a partitioned dataset? I assume that this would be the `_metadata` file?

If your search for *"Assuming that a single-file metadata solution is currently missing"* in the notebook I referenced, you will see where I am unsure of the best way to write/read metadata to/from a centralized location using `pyarrow.parquet`.  Let me know if you have any recommendations :)

**Comment by wesm** [comment]
Can you write to one of the Arrow mailing lists about this?

**Comment by jorisvandenbossche** [comment]
@rjzamora the writing functionality of combined metadata is indeed still missing, as you correctly stated. See https://issues.apache.org/jira/browse/ARROW-1983 for that. And I also just opened https://issues.apache.org/jira/browse/ARROW-5349 for the file path of metadata (another piece of the puzzle to get this dataset metadata working, I think).

But it would still be good to have a mailing list discussion about I think, to get a better picture of the functionalities that are still needed here.

**Comment by rjzamora** [comment]
Thanks for pointing me to [Arrow-5349](https://issues.apache.org/jira/browse/ARROW-5349) @jorisvandenbossche! I do think it's probably an important piece of the puzzle.

Also, I just posted my question to the mailing list - Thanks!

**Pull request closed without merging**